### PR TITLE
[DataLoader] Add simple `groupby` DataPipe

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* #60679 [DataLoader] Add Example Only `fork` DataPipe
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* **#60675 [DataLoader] Add simple `groupby` DataPipe**

Differential Revision: [D29461082](https://our.internmc.facebook.com/intern/diff/D29461082)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Entire stack is cleaner attempt to merge/rebase #59053

File: torch/utils/data/datapipes/iter/grouping.py [hunk 1]
```diff
@@ -205,6 +205,84 @@ def cmp_fn(a : Tuple[str, Any], b : Tuple[str, Any]):
...
+            for key in buffer.keys():
+                if len(buffer[key]) > biggest_size:
+                    biggest_size = len(buffer[key])
+                    biggest_key = key
+
+            if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                raise Exception('Failed to group items', str(buffer[biggest_key]))
+
+            if self.guaranteed_group_size is None or biggest_size >= self.guaranteed_group_size:
+                yield buffer[biggest_key]
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:280]: It seems we can have a function to return `biggest_key` to reduce the duplicated code.

File: torch/utils/data/datapipes/iter/grouping.py [hunk 2]
```diff
@@ -205,6 +205,84 @@ def cmp_fn(a : Tuple[str, Any], b : Tuple[str, Any]):
...
+            if buffer_size == self.buffer_size:
+                biggest_key = None
+                biggest_size = 0
+                for findkey in buffer.keys():
+                    if len(buffer[findkey]) > biggest_size:
+                        biggest_size = len(buffer[findkey])
+                        biggest_key = findkey
+
+                if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                    raise Exception('Failed to group items', str(buffer[biggest_key]))
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:249]: nit: `RuntimeError` here?

**Review by ejguan — APPROVED** [review]
LGTM with a few nit comments.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

**Pull request closed without merging**